### PR TITLE
Refactor DETR model to pl lightning

### PR DESCRIPTION
**Pull request by yangtcai**
#26

### CODE REVIEW

**Review by williamstark01 — CHANGES_REQUESTED** [review]
Very nice work!

Could you check the imports in train.py, it crashes when I run:
```
python train.py --configuration configuration.yaml
```

**Comment by yangtcai** [comment]
Maybe this time it can be trained :D

**Review by williamstark01 — CHANGES_REQUESTED** [review]
Works great now!

Could you update the log directory path in `submit_cluster_job.py` so that all log files for each experiment are saved in the same directory?

(You will also need to pass the datetime as an argument to the pipeline script:
https://github.com/Ensembl/gene_pcp/blob/main/submit_LSF_job.py#L88
https://github.com/Ensembl/gene_pcp/blob/main/transformer_pipeline.py#L386
https://github.com/Ensembl/gene_pcp/blob/main/transformer_pipeline.py#L416

**Comment by williamstark01** [comment]
Hey, you deleted the wrong datetime instantiation, can you take a look again? (Give it a test before you commit the code to make sure it works correctly.)

**Review by williamstark01 — CHANGES_REQUESTED** [review]
Hey, you deleted the wrong datetime instantiation, can you take a look again? (Give it a test before you commit the code to make sure it works correctly.)

**Comment by yangtcai** [comment]
Hi, @williamstark01, I have found what's happening in implementing Lightning, it seems like https://github.com/Ensembl/gene_pcp/blob/121be9895d414da3f13b5c8ec7588754e03336e1/transformer_pipeline.py#L468 will lead to the problem, and I test the callback function when in my friend's cluster, the same code will stop at 14 epochs compared our 5 or 6 epochs. I'm not sure the callback fuction is necessary for our project, what do you think about it?

**Comment by williamstark01** [comment]
The last commit with the new mAP is large and difficult to see what changed. Could you add a short description on what was updated and why?

**Comment by yangtcai** [comment]
Thanks for telling me about the `Early stopping`, I will add this as soon as possible.
The commit about the mAP is mainly because I need an identifier of the sequence. When we compute the mAP, we will choose the prediction which has the greatest IOU with the ground truth. When we do the matching, there is a situation we need to avoid that the prediction in the sample1 matches the ground truth in sample2. So I use `seq_start` as the identifier,  in dataloader stage, It will produce the sequence itself and its `seq_start`, in the training stage, the prediction value also contains the `seq_start` and the original prediction info (the class and the position).

File: configuration.yaml [hunk 1]
```diff
@@ -37,12 +37,12 @@ nhead: 8
 lr: 0.0001
 max_epochs: 300
 batch_size: 2
-validation_split: 0.2 
+separate_split: 0.3
```
Comment (williamstark01) [inline configuration.yaml:40]: Probably a better name for the `separate_split` parameter is something like `validation_test_ratio`.

File: dataloader.py [hunk 2]
```diff
@@ -225,14 +230,18 @@ def build_dataloader(configuration):
...
+    split = int(np.floor(separate_split * dataset_size))
 
     np.random.seed(configuration.seed)
     np.random.shuffle(indices)
-    train_indices, val_indices = indices[split:], indices[:split]
+    train_indices, val_indices, test_indices = (
+        indices[split:],
+        indices[:split],
+        indices[split:],
+    )
```
Comment (williamstark01) [inline dataloader.py:241]: The same indices are used here for the training and test sets, which makes them identical. We want to avoid this, it's important that the test set is completely distinct and separate from the training and validation sets.

**Review by williamstark01 — CHANGES_REQUESTED** [review]
Very good, you are splitting the dataset to three sets as well. There is a small bug in generating the indices, but great job otherwise.

I'm also getting these warnings during training:

```
[...]/mAP_validation.py:94: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  torch.tensor(detection[2]),
[...]/mAP_validation.py:95: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  torch.tensor(gt[2]),
```
A bug may be hiding there, it's probably safer to follow torch's recommendation and use their method for the tensor.

**Comment by williamstark01** [comment]
I noticed another warning:
```
UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 1. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`.
```
`batch_size` is equal to 2 in the config, so you may need to use the extra argument to self.log, or maybe something is wrong there?

File: mAP_validation.py [hunk 3]
```diff
@@ -9,81 +12,106 @@ def segment_IOU(segment1, segment2):
...
-    """
-    Calculates mean average precision
-    Parameters:
-        outputs: dictionary from dataloader
-        targets: Similar as outputs
-        iou_threshold (float): threshold where predicted overlap is correct
-        num_classes (int): number of classes
-    Returns:
-        float: mAP value across all classes given a specific IoU threshold
-    """
```
Comment (williamstark01) [inline mAP_validation.py:25]: I guess this was deleted by mistake?

File: mAP_validation.py [hunk 4]
```diff
@@ -9,81 +12,106 @@ def segment_IOU(segment1, segment2):
...
-    # list storing all AP for respective classes
+
+    # pred_logits = torch.rand(batch_size, num_queries, num_classes + 1)
+    # pred_boxes = torch.rand(batch_size, num_queries, 2)
+    # outputs = {"pred_logits": pred_logits, "pred_boundaries": pred_boxes}
+
+    #  {
+    #         "classes": torch.randint(low=0, high=num_classes, size=(num_target_boxes,)),
+    #         "coordinates": torch.rand(num_target_boxes, 2),
+    #     }
```
Comment (williamstark01) [inline mAP_validation.py:27]: We can delete this if not used?

File: dataloader.py [hunk 5]
```diff
@@ -225,14 +229,19 @@ def build_dataloader(configuration):
...
+    split1 = int(np.floor(configuration.validation_ratio * dataset_size))  # 0.6
+    split2 = int(np.floor(configuration.test_ratio * dataset_size))  # 0.2
     np.random.seed(configuration.seed)
     np.random.shuffle(indices)
-    train_indices, val_indices = indices[split:], indices[:split]
+    train_indices, val_indices, test_indices = (
+        indices[0:split1],
+        indices[split1:split2],
+        indices[split2:],
+    )
```
Comment (williamstark01) [inline dataloader.py:241]: This is tricky to understand, maybe the following is a better option?
```
    validation_size = int(configuration.validation_ratio * dataset_size)
    test_size = int(configuration.test_ratio * dataset_size)
    np.random.seed(configuration.seed)
    np.random.shuffle(indices)
    val_indices, test_indices, train_indices = (
        indices[:validation_size],
        indices[validation_size:validation_size+test_size],
        indices[validation_size+test_size:],
    )
```
(And reduce the validation and test ratio values in the config file accordingly.)

**Review by williamstark01 — CHANGES_REQUESTED** [review]
A couple of minor edits and it's ready to merge!

File: model.py [hunk 6]
```diff
@@ -221,6 +179,157 @@ def forward(self, outputs, targets):
...
+        out = {
+            "pred_logits": outputs_class,
+            "pred_boundaries": outputs_coord,
+            "seq_start": seq_starts,
+        }
+        return out
+
+    def training_step(self, batch, batch_idx):
+        samples, seq_starts, targets = batch
+        targets = [{k: v for k, v in t.items()} for t in targets]
```
Comment (williamstark01) [inline model.py:268]:
```
        targets = [{k: v for k, v in t.items()} for t in targets]
```
What does this line do? It seems like it is not changing `targets` at all to me.

File: train.py [hunk 7]
```diff
@@ -4,105 +4,55 @@
...
-    writer,
-):
-    losssum = 0
-    sample_num = 0
-    model.eval()
-    criterion.eval()
-    with torch.no_grad():
-        for samples, targets in data_loader:
-            samples = samples.to(device)
-            targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
```
Comment (yangtcai) [inline train.py:76]: Hi, @williamstark01, in the former version, I need to add this command to move our datasets to GPU, however, when we use pl, it seems not useful. I will test if it delete useful.

File: model.py (hunk 6 above)
Comment (yangtcai) [reply model.py]: The test result seems useless, and now I remove it.

**Review by williamstark01 — APPROVED** [review]
Excellent work, one of the most extensive PRs to merge so far!

closes #26 
closes #44
